torch/csrc/DynamicTypes.cpp

```diff
@@ -117,18 +117,74 @@ PyObject* createPyObject(
   return obj.release();
 }
 
+PyTypeObject* loadTypedStorageTypeObject() {
+  PyObject* storage_module = PyImport_ImportModule("torch.storage");
+  TORCH_INTERNAL_ASSERT(storage_module && PyModule_Check(storage_module));
+
+  PyObject* typed_storage_obj = PyObject_GetAttrString(storage_module, "TypedStorage");
+  TORCH_INTERNAL_ASSERT(typed_storage_obj && PyType_Check(typed_storage_obj));
+  return reinterpret_cast<PyTypeObject*>(
+      PyObject_GetAttrString(storage_module, "TypedStorage"));
+}
+
+PyTypeObject* getTypedStorageTypeObject() {
+  // NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+  static PyTypeObject* typed_storage_type_obj = loadTypedStorageTypeObject();
+  return typed_storage_type_obj;
+}
+
 bool isStorage(PyObject* obj)
 {
-  return py_storage_type_to_attype.count(Py_TYPE(obj));
+  if (PyObject_TypeCheck(obj, getTypedStorageTypeObject())) {
+    return true;
+  }
+  auto obj_type = Py_TYPE(obj);
+  for (auto const& item : py_storage_type_to_attype) {
+    auto const& storage_type = item.first;
+    if (obj_type == storage_type) {
+      return true;
+    }
+  }
+  return false;
 }
-at::Storage createStorage(PyObject* obj)
+
+at::Storage createStorageGetType(PyObject* obj, at::ScalarType& scalar_type)
 {
-  auto it = py_storage_type_to_attype.find(Py_TYPE(obj));
-  if (it == py_storage_type_to_attype.end()) {
-    throw TypeError("not a storage '%s'", Py_TYPE(obj)->tp_name);
+  bool is_typed_storage = PyObject_TypeCheck(obj, getTypedStorageTypeObject());
+  THPObjectPtr maybe_untyped_storage;
+  if (is_typed_storage) {
+    PyObject* maybe_untyped_storage_obj = PyObject_GetAttrString(obj, "_storage");
+    TORCH_INTERNAL_ASSERT(maybe_untyped_storage_obj);
+    maybe_untyped_storage = maybe_untyped_storage_obj;
   }
-  auto& type = *it->second;
-  return type.unsafeStorageFromTH(((THPVoidStorage*)obj)->cdata, true);
+
+  auto obj_type = Py_TYPE(obj);
+  for (auto const& item : py_storage_type_to_attype) {
+    auto const& storage_type = item.first;
+    if (is_typed_storage) {
+      if (Py_TYPE(maybe_untyped_storage.get()) == storage_type) {
```

Comment:
I'm not sure why you have to test `storage_type` here. You don't end up using it below; you just read out the dtype directly from the dtype field. This will also make it easier to dump the FloatStorage classes later.

Reply:
Yeah, this function is confusing. `storage_type` resolves to either `torch.ByteStorage` or `torch.cuda.ByteStorage` (if the CUDA one has been registered yet), not the scalar type. So I think `storage_type` still needs to be tested, but we could simplify things. We shouldn't need the `py_storage_type_to_attype` map anymore--we can just do a simple if statement. I can make that change in this PR if you'd like